server/src/main/java/org/opensearch/common/cache/stats/MultiDimensionCacheStats.java

```diff
@@ -0,0 +1,160 @@
+ */
+
+package org.opensearch.common.cache.stats;
+
+import org.opensearch.core.common.io.stream.StreamInput;
+import org.opensearch.core.common.io.stream.StreamOutput;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Map;
+import java.util.TreeMap;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * A CacheStats object supporting aggregation over multiple different dimensions.
+ * Stores a fixed snapshot of a cache's stats; does not allow changes.
+ */
+public class MultiDimensionCacheStats implements CacheStats {
+    // A snapshot of a StatsHolder containing stats maintained by the cache.
+    // Pkg-private for testing.
+    final Map<StatsHolder.Key, CacheStatsResponse.Snapshot> snapshot;
+    final List<String> dimensionNames;
+
+    public MultiDimensionCacheStats(Map<StatsHolder.Key, CacheStatsResponse.Snapshot> snapshot, List<String> dimensionNames) {
+        this.snapshot = snapshot;
+        this.dimensionNames = dimensionNames;
+    }
+
+    public MultiDimensionCacheStats(StreamInput in) throws IOException {
+        this.dimensionNames = List.of(in.readStringArray());
+        Map<StatsHolder.Key, CacheStatsResponse.Snapshot> readMap = in.readMap(
+            i -> new StatsHolder.Key(List.of(i.readArray(StreamInput::readString, String[]::new))),
+            CacheStatsResponse.Snapshot::new
+        );
+        this.snapshot = new ConcurrentHashMap<StatsHolder.Key, CacheStatsResponse.Snapshot>(readMap);
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        out.writeStringArray(dimensionNames.toArray(new String[0]));
+        out.writeMap(
+            snapshot,
+            (o, key) -> o.writeArray((o1, dimValue) -> o1.writeString((String) dimValue), key.dimensionValues.toArray()),
+            (o, snapshot) -> snapshot.writeTo(o)
+        );
+    }
+
+    @Override
+    public CacheStatsResponse.Snapshot getTotalStats() {
+        CacheStatsResponse response = new CacheStatsResponse();
+        // To avoid making many Snapshot objects for the incremental sums, add to a mutable CacheStatsResponse and finally convert to
+        // Snapshot
+        for (Map.Entry<StatsHolder.Key, CacheStatsResponse.Snapshot> entry : snapshot.entrySet()) {
+            response.add(entry.getValue());
+        }
+        return response.snapshot();
+    }
+
+    @Override
+    public long getTotalHits() {
+        return getTotalStats().getHits();
+    }
+
+    @Override
+    public long getTotalMisses() {
+        return getTotalStats().getMisses();
+    }
+
+    @Override
+    public long getTotalEvictions() {
+        return getTotalStats().getEvictions();
+    }
+
+    @Override
+    public long getTotalSizeInBytes() {
+        return getTotalStats().getSizeInBytes();
+    }
+
+    @Override
+    public long getTotalEntries() {
+        return getTotalStats().getEntries();
+    }
```

Comment:
They were originally not expensive because we tracked total stats, to spare having to do these additions. I suppose we can remove them, but they are useful in tests.